### PR TITLE
Dispute spam protection

### PR DESCRIPTION
**Pull request by eskimor**
# Status:

- [x] Everything important is implemented
- [x] Fix existing tests
- [x] Write tests

Also missing, but will be a separate PR: Guide changes.

skip check-dependent-cumulus

### CODE REVIEW

File: node/core/approval-voting/src/import.rs [hunk 1]
```diff
@@ -76,7 +76,7 @@ struct ImportedBlockInfo {
 }
 
 struct ImportedBlockInfoEnv<'a> {
-	session_window: &'a RollingSessionWindow,
+	session_window: &'a Option<RollingSessionWindow>,
```
Comment (rphmeier) [inline node/core/approval-voting/src/import.rs:79]: seems really out of scope

File: roadmap/implementers-guide/src/node/disputes/dispute-coordinator.md [hunk 2]
```diff
@@ -75,8 +74,7 @@ struct State {
 
 Check DB for recorded votes for non concluded disputes we have not yet
 recorded a local statement for.
-For all of those send `DisputeParticipationMessage::Participate` message to
-dispute participation subsystem.
+For all of those initiate dispute participation.
```
Comment (rphmeier) [inline roadmap/implementers-guide/src/node/disputes/dispute-coordinator.md:77]: Dispute participation was deleted. But the changeset here doesn't describe what 'initiating dispute participation' even means.

File: node/primitives/src/lib.rs [hunk 3]
```diff
@@ -82,6 +76,58 @@ pub const BACKING_EXECUTION_TIMEOUT: Duration = Duration::from_secs(2);
...
+	/// Get the value as `SessionIndex` for doing comparisons with those.
+	pub fn get(self) -> SessionIndex {
+		self.0
+	}
+
+	/// Helper function for `new_session_window_size`.
+	///
+	/// Don't use it. The only reason it is public, is because otherwise the
+	/// `new_session_window_size` macro would not work outside of this module.
+	pub const fn unchecked_new(size: SessionIndex) -> Self {
```
Comment (rphmeier) [inline node/primitives/src/lib.rs:126]: Add `doc(hidden)`, then

File: node/primitives/src/lib.rs [hunk 4]
```diff
@@ -82,6 +76,58 @@ pub const BACKING_EXECUTION_TIMEOUT: Duration = Duration::from_secs(2);
...
+///
+/// See: https://github.com/rust-lang/rust/issues/67441
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct SessionWindowSize(SessionIndex);
+
+#[macro_export]
+/// Create a new checked `SessionWindowSize`
+///
+/// which cannot be 0.
+macro_rules! new_session_window_size {
```
Comment (rphmeier) [inline node/primitives/src/lib.rs:92]: This goes is the opposite direction that we'd like, btw, which is that we draw the session window size from the on-chain logic.

File: node/core/dispute-coordinator/src/real/error.rs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use futures::channel::oneshot;
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/error.rs:17]: Missing module docs

File: node/core/dispute-coordinator/src/real/ordering.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+///
+/// Note: That by `oldest` we mean oldest in terms of relay chain block number, for any block
+/// number that has not yet been finalized. If a block has been finalized already it should be
+/// treated as low priority when it comes to disputes, as even in the case of a negative outcome,
+/// we are already too late. The ordering mechanism here serves to prevent this from happening in
+/// the first place.
+#[derive(Copy, Clone)]
+pub struct CandidateComparator {
+	/// Relay chain block number the candidate got included in.
+	included_block_number: BlockNumber,
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/ordering.rs:57]: As discussed, this is not unique. relay-parent number is, though.

File: node/core/dispute-coordinator/src/real/ordering.rs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	/// Para id, used for ordering parachains within same relay chain block.
+	para_id: Id,
+	/// The hash of the relay chain block the candidate got included in.
+	included_block_hash: Hash,
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/ordering.rs:61]: As discussed, this is not unique.

File: node/core/dispute-coordinator/src/real/ordering.rs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+/// suffice for this goal, ordering by age ensures we are not only resolving disputes, but also
+/// resolve the oldest one first, which are also the most urgent and important ones to resolve.
+///
+/// Note: That by `oldest` we mean oldest in terms of relay chain block number, for any block
+/// number that has not yet been finalized. If a block has been finalized already it should be
+/// treated as low priority when it comes to disputes, as even in the case of a negative outcome,
+/// we are already too late. The ordering mechanism here serves to prevent this from happening in
+/// the first place.
+#[derive(Copy, Clone)]
+pub struct CandidateComparator {
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/ordering.rs:55]: In the other two comments I suggest using `relay_parent_number` and `relay_parent_hash` instead of `included_block_X`. But `relay_parent ++ para_id` is also not unique since multiple collators can build a parachain block. I suggest `(relay_parent_number, candidate_hash)` as the comparator.

File: node/core/dispute-coordinator/src/real/ordering.rs [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+/// Objective ordering between nodes is important in case of lots disputes, so nodes will pull in
+/// the same direction and work on resolving the same disputes first. This ensures that we will
+/// conclude some disputes, even if there are lots of them. While any objective ordering would
+/// suffice for this goal, ordering by age ensures we are not only resolving disputes, but also
+/// resolve the oldest one first, which are also the most urgent and important ones to resolve.
+///
+/// Note: That by `oldest` we mean oldest in terms of relay chain block number, for any block
+/// number that has not yet been finalized. If a block has been finalized already it should be
+/// treated as low priority when it comes to disputes, as even in the case of a negative outcome,
+/// we are already too late. The ordering mechanism here serves to prevent this from happening in
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/ordering.rs:52]: 'too late' vs. 'way too late' is an important distinction, but I do see and understand that we don't want wasteful disputes on finalized candidates to crowd out other disputes.

File: node/core/dispute-coordinator/src/real/participation/mod.rs [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/participation/mod.rs:16]: Module docs please!

File: node/core/dispute-coordinator/src/real/participation/mod.rs [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+mod queues;
+use queues::Queues;
+pub use queues::{Error as QueueError, ParticipationRequest};
+
+/// How many participation processes do we want to run in parallel the most.
+///
+/// This should be a relatively low value, while we might have a speedup once we fetched the data,
+/// due to multi core architectures, but the fetching itself can not be improved by parallel
+/// requests. This means that higher numbers make it harder for a single dispute to resolve fast.
+const MAX_PARALLEL_PARTICIPATIONS: usize = 3;
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/participation/mod.rs:48]: Follow-up issue: make this configurable?

File: node/core/dispute-coordinator/src/real/participation/queues.rs [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+			}
+			// Remove any best effort entry:
+			self.best_effort.remove(&req.candidate_hash);
+			self.priority.insert(comparator, req);
+		} else {
+			if self.best_effort.len() >= BEST_EFFORT_QUEUE_SIZE {
+				return Err(Error::BestEffortFull)
+			}
+			self.best_effort
+				.entry(req.candidate_hash)
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/participation/queues.rs:143]: `queue(None, req)` can lead to something being present in both best effort and participation queues.

File: node/primitives/src/lib.rs (hunk 4 above)
Comment (eskimor) [reply node/primitives/src/lib.rs:92]: That actually should be fine - in that case it would just be a runtime check instead of a compile time check.

File: node/core/dispute-coordinator/src/real/participation/mod.rs [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+
+async fn participate(
+	mut result_sender: WorkerMessageSender,
+	mut sender: impl SubsystemSender,
+	block_hash: Hash,
+	req: ParticipationRequest,
+) {
+	// in order to validate a candidate we need to start by recovering the
+	// available data
+	let (recover_available_data_tx, recover_available_data_rx) = oneshot::channel();
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/participation/mod.rs:246]: The participation logic should start with some checks about whether we need to participate. We shouldn't participate in any disputes that we already voted in. Or is this protected somewhere else?

File: node/core/dispute-coordinator/src/real/spam_slots.rs [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+		candidate: CandidateHash,
+		validator: ValidatorIndex,
+	) -> bool {
+		let c = self.slots.entry((session, validator)).or_default();
+		if *c >= MAX_SPAM_VOTES {
+			return false
+		}
+		*c += 1;
+		let validators = self.unconfirmed.entry((session, candidate)).or_default();
+		validators.insert(validator);
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/spam_slots.rs:92]: We should only increment `c` and return false if this is `false`. Otherwise, calling `add_unconfirmed` twice with the same candidate&validator will increment spam slots twice.

File: node/core/dispute-coordinator/src/real/mod.rs [hunk 15]
```diff
@@ -112,12 +117,36 @@ impl Config {
...
-	store: Arc<dyn KeyValueDB>,
-	keystore: Arc<LocalKeystore>,
-	metrics: Metrics,
+impl<Context> overseer::Subsystem<Context, SubsystemError> for DisputeCoordinatorSubsystem
+where
+	Context: SubsystemContext<Message = DisputeCoordinatorMessage>,
+	Context: overseer::SubsystemContext<Message = DisputeCoordinatorMessage>,
+{
+	fn start(self, ctx: Context) -> SpawnedSubsystem {
+		let future = async {
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/mod.rs:126]: nice rewrite.

File: node/core/dispute-coordinator/src/real/mod.rs [hunk 16]
```diff
@@ -128,1159 +157,225 @@ impl DisputeCoordinatorSubsystem {
...
-		Some(votes) => votes,
-		None =>
-			if let MaybeCandidateReceipt::Provides(candidate_receipt) = candidate_receipt {
-				CandidateVotes { candidate_receipt, valid: Vec::new(), invalid: Vec::new() }
-			} else {
+			// Note: It seems we really should not receive any messages before the first
+			// `ActiveLeavesUpdate`, if that proves wrong over time and we do receive
+			// messages before the first `ActiveLeavesUpdate` that should not be dropped,
+			// this can easily be fixed by collecting those messages and passing them on to
+			// `Initialized::new()`.
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/mod.rs:373]: +1...still, a bit sketchy we don't do this already.

File: node/core/dispute-coordinator/src/real/initialized.rs [hunk 17]
```diff
@@ -0,0 +1,1104 @@
...
+	highest_session: SessionIndex,
+	spam_slots: SpamSlots,
+	participation: Participation,
+	ordering_provider: OrderingProvider,
+	participation_receiver: WorkerMessageReceiver,
+	metrics: Metrics,
+}
+
+impl Initialized {
+	pub fn new(
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/initialized.rs:82]: missing docs

File: node/core/dispute-coordinator/src/real/initialized.rs [hunk 18]
```diff
@@ -0,0 +1,1104 @@
...
+			highest_session,
+			spam_slots,
+			ordering_provider,
+			participation,
+			participation_receiver,
+			metrics,
+		}
+	}
+
+	pub async fn run<B, Context>(
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/initialized.rs:106]: missing docs

File: node/core/dispute-coordinator/src/real/initialized.rs [hunk 19]
```diff
@@ -0,0 +1,1104 @@
...
+		backend: &mut B,
+		participations: &mut Vec<(Option<CandidateComparator>, ParticipationRequest)>,
+		clock: &dyn Clock,
+	) -> Result<()>
+	where
+		Context: overseer::SubsystemContext<Message = DisputeCoordinatorMessage>,
+		Context: SubsystemContext<Message = DisputeCoordinatorMessage>,
+		B: Backend,
+	{
+		for (comparator, request) in std::mem::take(participations).into_iter() {
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/initialized.rs:146]: `participations.drain(..)`

File: node/core/dispute-coordinator/src/real/participation/queues.rs [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+	/// This mechanism leads to an amplifying effect - the more validators already participated,
+	/// the more likely it becomes that more validators will participate soon, which should lead to
+	/// a quick resolution of disputes, even in the best effort queue.
+	best_effort: HashMap<CandidateHash, BestEffortEntry>,
+
+	/// Priority queue.
+	///
+	/// In the priority queue, we have a strict ordering of candidates and participation will
+	/// happen in that order.
+	priority: BTreeMap<CandidateComparator, ParticipationRequest>,
```
Comment (ordian) [inline node/core/dispute-coordinator/src/real/participation/queues.rs:57]: why not use https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html?

File: node/core/dispute-coordinator/src/real/participation/mod.rs (hunk 13 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/real/participation/mod.rs:246]: That's the same as before - checked on import. Participation will make sure though that no participation gets queued, if we are already participating at that moment. What could happen that approval voting casts a vote, after we already queued participation - in that case we would in fact duplicate work, but nothing bad should happen as on `issue_local_statement` we will only cast votes for disputes we did not yet have cast a vote for.

It seems that the risk of duplicate work in this race condition should not matter much, so I would refrain complicating things by adding cancel logic for that case.

File: node/core/dispute-coordinator/src/real/spam_slots.rs (hunk 14 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/real/spam_slots.rs:92]: Good point. Multiple identical imports are also spam, but yeah - pure flood should already be handled properly by distribution.

File: node/core/dispute-coordinator/src/real/participation/queues.rs (hunk 12 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/real/participation/queues.rs:143]: Good catch!

File: node/core/dispute-coordinator/src/real/participation/mod.rs (hunk 10 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/real/participation/mod.rs:16]: It's in real/mod.rs already. When you start reading at the top, you get an immediate overview and don't have to jump into the module if you are not interested in it right now.

File: node/core/dispute-coordinator/src/real/participation/queues.rs (hunk 20 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/real/participation/queues.rs:57]: Mostly interface. `BinaryHeap` has a set like interface, while I need a map here, other than that `BinaryHeap` would be better here.

File: node/core/dispute-coordinator/src/real/ordering.rs [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+
+impl Ord for CandidateComparator {
+	fn cmp(&self, other: &Self) -> Ordering {
+		match self.relay_parent_block_number.cmp(&other.relay_parent_block_number) {
+			Ordering::Equal => (),
+			o => return o,
+		}
+		self.candidate_hash.cmp(&other.candidate_hash)
+	}
+}
```
Comment (ordian) [inline node/core/dispute-coordinator/src/real/ordering.rs:102]: why not derive this?

File: node/core/dispute-coordinator/src/real/participation/queues.rs (hunk 20 above)
Comment (ordian) [reply node/core/dispute-coordinator/src/real/participation/queues.rs:57]: But it is used as a set, because there is no lookup on `CandidateComparator`? So it's possible to create a wrapper
`OrderedParticipationRequest` which is a tuple with a comparison on the first field. And store it in the `BinaryHeap`.

Anyway, this is minor impl detail.

Comment (eskimor) [reply node/core/dispute-coordinator/src/real/participation/queues.rs:57]: Yeah, that's what I meant. Don't think the added code complexity is worth it here - disputes are not performance critical, they should just work if they happen.

File: node/core/dispute-coordinator/src/real/ordering.rs (hunk 21 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/real/ordering.rs:102]: Given that it is critical how the comparing is done - I preferred being explicit.

File: node/core/dispute-coordinator/src/real/participation/queues/tests.rs [hunk 22]
```diff
@@ -0,0 +1,131 @@
+use assert_matches::assert_matches;
```
Comment (rphmeier) [inline node/core/dispute-coordinator/src/real/participation/queues/tests.rs:1]: imports are floating above license header

**Review by rphmeier — APPROVED** [review]
LGTM. What's left?

File: node/core/dispute-coordinator/src/real/participation/queues/tests.rs (hunk 22 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/real/participation/queues/tests.rs:1]: haha, thanks - noticed that as well. The beauty of auto import. :smile:

**Comment by eskimor** [comment]
> LGTM. What's left?

Want to add a few more tests, other than that it should be good. Doc improvements and such don't need to block this PR, but can be follow ups.

File: node/core/dispute-coordinator/src/real/participation/mod.rs (hunk 11 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/real/participation/mod.rs:53]: Will file one.

**Comment by eskimor** [comment]
cumulus companion: https://github.com/paritytech/cumulus/pull/765

**Review by ordian — APPROVED** [review]

**Comment by ordian** [comment]
bot merge
